Zapier Process-Action Feature

aggroskater opened the pull request:
# Implements #1983 

## Description

Adds support for pushing processAction events to Zapier (or any other endpoint) and introduces a configuration object `ZAPIER_CONFIG_OBJECT` for specifying which Zapier webhook URLs should be associated with which answers/responses. Also introduces a timeout config option `ZAPIER_TIMEOUT_MS` for the endpoint.

# Checklist:

- [X] I have manually tested my changes on desktop and mobile
- [X] The test suite passes locally with my changes
- [X] If my change is a UI change, I have attached a screenshot to the description section of this pull request
- [X] [My change is 300 lines of code or less](https://github.com/MoveOnOrg/Spoke/blob/main/CONTRIBUTING.md#your-first-code-contribution), or has a documented reason in the description why it’s longer
- [X] I have made any necessary changes to the documentation
- [X] I have added tests that prove my fix is effective or that my feature works
- [X] My PR is labeled [WIP] if it is in progress

schuyler1d reviewed (changes requested):
This looks good.  Can you remove the commented out`log.info`s and big switch statement and rebase so it only makes changes to the files that have final differences?  Happy to help you with that if you're unfamilar with git-rebasing.

aggroskater commented:
I've cleaned up the commented out code now. It's mostly just the action handler in `src/extensions/action-handlers/zapier-action.js`, but it also looks like I added a parameter in `src/workers/tasks.js` at some point. It doesn't look like I'm actually using it anywhere though. I'll poke around a bit and see if that change can be reverted too.

It looks like y'all're working on an 11.0 branch (`stage-main-11-0`) but it hasn't landed in the `main` branch yet. I can rebase the PR on main once 11.0 lands, I suppose. I'll also work on getting some unit tests for the extension, and updating the documentation to explain the expected format of `ZAPIER_CONFIG_OBJECT`.

schuyler1d commented on src/extensions/action-handlers/zapier-action.js:
```diff
@@ -71,9 +71,80 @@ export async function onTagUpdate(
+      interactionStep: interactionStep,
+      campaignContactId: campaignContactId,
+      contact: contact,
+      campaign: campaign,
```
it will be rare, but there could be credential-y info in `campaign.features` object, so maybe in an abundance of caution let's do:
```
const campaignCopy = {...campaign};
delete campaignCopy.feature;
delete campaignCopy.features;
const payload = {
   questionResponse
   interactionStep,
   campaignContactId,
   contact,
   campaign: campaignCopy
....
   };
```
(note that in es6 you don't need to stutter the values that have the same name, so clean that up here, as well)

schuyler1d commented on src/extensions/action-handlers/zapier-action.js:
```diff
@@ -71,9 +71,80 @@ export async function onTagUpdate(
+    var final_url = "";
+    const answer_option = interactionStep.answer_option;
+
+    const config = JSON.parse(getConfig("ZAPIER_CONFIG_OBJECT", organization)) || {};
```
this will fail when ZAPIER_CONFIG_OBJECT doesn't exist so let's do:
`const config = JSON.parse(getConfig("ZAPIER_CONFIG_OBJECT", organization) || "{}");`

schuyler1d commented on src/extensions/action-handlers/zapier-action.js:
```diff
@@ -71,9 +71,80 @@ export async function onTagUpdate(
+    const zap_timeout = getConfig("ZAPIER_TIMEOUT_MS", organization) || 5000;
+    log.info(`Zapier timeout: ${zap_timeout}`);
+
+    var final_url = "";
```
use `let finalUrl = "";` for style purposes

schuyler1d reviewed (changes requested):
This looks good modulo the changes.

One thing I'm not sure of -- for people using the old version of this that get the pings on the tags only, is this something maybe we should have a second env var (instead of re-using: ZAPIER_WEBHOOK_URL -- maybe just ZAPIER_ACTION_URL)  and then if neither ZAPIER_ACTION_URL nor ZAPIER_CONFIG_OBJECT exist we bail early (at the top of process action)

aggroskater commented:
@schuyler1d Thanks for the feedback :) I'm working on getting my development environment up and running so that I can implement your changes and contribute some tests for the action-handler. I've also got a note to update the environment variable reference.

I'm giving myself a public deadline here: have this PR ready to go by the end of next weekend at the latest (by 2021-08-22).

aggroskater commented:
Note to self: the local test harness *is* passing against the docker-compose environment, with 8 tests skipped, 582 passed, 590 total (off of my zapier-test branch). But I need to make changes first before ticking that "tests pass locally" box.

aggroskater commented:
@schuyler1d I have added a test suite for the integration and added documentation that is (hopefully) sufficiently detailed to get folks started. I believe I also added the other small changes (like introducing `ZAPIER_ACTION_URL`). Let me know if there are any other changes needed, or if this is good for merging (or rebasing onto another branch and then merging).

schuyler1d commented on __test__/extensions/action-handlers/zapier-action.test.js:
```diff
@@ -0,0 +1,884 @@
+        external_id: "1",
+        first_name: "Preston",
+        last_name: "Maness",
+        cell: "+15129551048",
```
Please change the cell to a fake one (regardless of whether this is a real number) -- if you would like your name in the tests, I'm fine with that, but I'd prefer an obviously fake name.
For fake phone numbers examples are in:
__test__/test_data/female_scientists.csv which you can grab from -- see https://www.businessinsider.com/555-phone-number-tv-movies-telephone-exchange-names-ghostbusters-2018-3 for why these are fake.

I don't want PII or even something that could be mistaken for PII in the project.

schuyler1d commented on docs/HOWTO_INTEGRATE_WITH_ZAPIER.md:
```diff
@@ -0,0 +1,419 @@
+    "external_id": "2",
+    "first_name": "Patrick",
+    "last_name": "Maness",
+    "cell": "+15129031315",
```
ditto

schuyler1d reviewed (commented):
one last small tweak -- great on adding tests and documentation!

aggroskater commented:
Roger that :) I've gone through and swapped out the test data. Confirmation that it's not there:

```
20:06:34 preston@neo:~/Spoke (zapier-test)$ grep -i -r 'preston' __test__/
20:06:39 preston@neo:~/Spoke (zapier-test)$ grep -i -r 'patrick' __test__/
20:06:43 preston@neo:~/Spoke (zapier-test)$ grep -i -r 'patrick' docs/
20:06:48 preston@neo:~/Spoke (zapier-test)$ grep -i -r 'preston' docs/
20:06:53 preston@neo:~/Spoke (zapier-test)$ grep -i -r '5129551048' __test__/
20:07:02 preston@neo:~/Spoke (zapier-test)$ grep -i -r '5129031315' __test__/
20:07:13 preston@neo:~/Spoke (zapier-test)$ grep -i -r '5129551048' docs/
20:07:17 preston@neo:~/Spoke (zapier-test)$ grep -i -r '5129031315' docs/
20:07:22 preston@neo:~/Spoke (zapier-test)$ 
```

Let me know whether or not to rebase, and if so, on which branch.

schuyler1d commented:
yeah, maybe a rebase would be a good idea (in this branch -- i think you can `push -f` and github will do the right thing (i.e. update this PR rather than delete it).  And then we're good to go!

aggroskater commented:
Alrighty. Rebased on latest `main` commit.

schuyler1d reviewed (approved):
:partying_face:
